### ISSUE TITLE
Many test cases fail on RHEL9 due to SHA1 not being supported for RSA signing

### ISSUE THREAD

**Issue by gsauthof** [ronf/asyncssh#566]
Many test cases fail on RHEL9 (Redhat  Enterprise Linux 9) like this:

```
======================================================================
ERROR: test_hostbased_auth (tests.test_auth._TestAuth) [Untrusted key]
Unit test host-based authentication
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/test_auth.py", line 527, in test_hostbased_auth
    await self.check_auth(b'hostbased', (False, None),
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/test_auth.py", line 416, in check_auth
    self.assertEqual((await client_conn.get_auth_result()),
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/test_auth.py", line 162, in get_auth_result
    return await self._auth_waiter
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/util.py", line 254, in _run_task
    await coro
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/auth.py", line 305, in _start
    await self.send_request(String(keypair.algorithm),
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/auth.py", line 136, in send_request
    await self._conn.send_userauth_request(self._method, *args, key=key,
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/test_auth.py", line 179, in send_userauth_request
    packet += String(key.sign(String('') + packet))
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/public_key.py", line 2253, in sign
    return self._key.sign(data, self.sig_algorithm)
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/public_key.py", line 564, in sign
    self.sign_ssh(data, sig_algorithm)))
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/rsa.py", line 242, in sign_ssh
    return String(self._key.sign(data, _hash_algs[sig_algorithm]))
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/crypto/rsa.py", line 134, in sign
    return priv_key.sign(data, PKCS1v15(), hashes[hash_name]())
  File "/usr/lib64/python3.9/site-packages/cryptography/hazmat/backends/openssl/rsa.py", line 501, in sign
    return _rsa_sig_sign(self._backend, padding, algorithm, self, data)
  File "/usr/lib64/python3.9/site-packages/cryptography/hazmat/backends/openssl/rsa.py", line 244, in _rsa_sig_sign
    pkey_ctx = _rsa_sig_setup(
  File "/usr/lib64/python3.9/site-packages/cryptography/hazmat/backends/openssl/rsa.py", line 213, in _rsa_sig_setup
    raise UnsupportedAlgorithm(
cryptography.exceptions.UnsupportedAlgorithm: sha1 is not supported by this backend for RSA signing.
```

or

```
======================================================================
ERROR: test_agent_forwarding (tests.test_channel._TestChannel)
Test SSH agent forwarding
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/util.py", line 94, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib64/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/test_channel.py", line 837, in test_agent_forwarding
    async with self.connect(username='ckey',
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/misc.py", line 274, in __aenter__
    self._coro_result = await self._coro
  File "/builddir/build/BUILD/asyncssh-2.13.1/tests/server.py", line 311, in connect
    return await asyncssh.connect(host or self._server_addr,
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/connection.py", line 8042, in connect
    return await asyncio.wait_for(
  File "/usr/lib64/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/builddir/build/BUILD/asyncssh-2.13.1/asyncssh/connection.py", line 440, in _connect
    await options.waiter
asyncssh.misc.ConnectionLost: Connection lost
```

Complete log: https://kojipkgs.fedoraproject.org//work/tasks/3956/100993956/build.log

Asyncssh version: 2.13.1

This reminds me of: #493

Background: This came up since a [few users are interested](https://bugzilla.redhat.com/show_bug.cgi?id=2196046_ to have asyncssh packaged in EPEL9 (basically this is the community repository for RHEL/RHEL-like distributions for packages that are available in Fedora but not in the RHEL main repository).

**Comment by ronf** [comment]
Thanks for the report!

I've been able to reproduce this here using the container setup mentioned in #493. However, unlike before where the errors were only applying when running OpenSSH (in interoperability tests), it looks like the errors are now happening when AsyncSSH calls into cryptography asking for signing with a SHA-1 hash. This could be a bit complicated to fix, as there are assumptions in a number of places that the default signature algorithm name is always the same as the key's algorithm name, but now I'll need to make sure I never use that for ssh-rsa keys, always specifying one of the alternatives such as rsa-sha2-256.

I've made some progress, but will have to experiment a bit to see what the cleanest approach is to provide a proper fix. I'll let you know when I have something.

**Comment by ronf** [comment]
An update on this: A large number of errors were triggered by using keys with OpenSSH certificates based on RSA. AsyncSSH didn't properly negotiate to use the available SHA-2 signature algorithms in this case, resulting in it using plain ssh-rsa, which relies on SHA-1 hashes. I have checked in code which adds support for this in commit 5be289e in the "develop" branch.

With this change, there are still 77 errors being reported when running the tests on RHEL9, but that's way down from 1373 in the original run. I'll be going through the remaining errors case by case. I expect many of them will be very easy to fix.

I'll report back when these are addressed.

**Comment by ronf** [comment]
I just submitted commit 72d9845 in the "develop" branch with changes to allow the unit tests to pass even on systems with SHA-1 signatures disabled. In some places, I switched over to ECDSA to avoid this issue, and in other places where the test involved selecting signature algorithms explicitly, I just had the tests ignore the UnsupportedAlgorithm error that gets raised on these systems.

The only remaining thing I'm seeing is what looks like a race condition generating the following output:

```
Unknown child process pid 28381, will report returncode 255
Loop <_UnixSelectorEventLoop running=False closed=True debug=False> that handles pid 28381 is closed
```

However, I don't think this is related to the SHA-1 issue on RHEL9. It seems like it might be a race condition within asyncio related to subprocesses from the little bit of digging I did. It appears to be harmless, and doesn't cause a test failure or other bad behavior, so I think this can be investigated separately. I actually don't see the error at all on macOS.

Let me know if you're still seeing any test failures with this change.

**Comment by ktdreyer** [comment]
I have run the unit tests on RHEL 9 today (`openssl-3.0.7-20.el9` and `python3-cryptography-36.0.1-4.el9`). The tests all pass now on `develop`. Thank you!

**Comment by ronf** [comment]
Great - thanks for letting me know! These changes will be included in the next release.

**Comment by ronf** [comment]
These changes are now available in AsyncSSH 2.13.2.
